adr/ADR-0001-trustyai-external-library-integration.md

```diff
@@ -0,0 +1,114 @@
+
+## Goals
+
+* Ability to extend TrustyAI with third-party libraries for explainability and fairness/bias metrics
+* Minimal impact on the current TrustyAI service external API
+* Must conform to Open Data Hub general architecture
+
+## Non-goals
+
+N/A
+
+## Current situation
+
+Since the TrustyAI algorithmic core is written in Java, as well as the TrustyAI service which provides all the ODH integration (ModelMesh payload consumption, data storage handling, Prometheus publishing and REST endpoints), the only way to add new explainers or metrics is to implement them in Java or add them as a dependency, if available as a Java library.
+Although the Python bindings for TrustyAI exist, their purpose is to invoke the Java algorithms from Python and not the other way around.
+At the moment there is no out of the box solution for TrustyAI to leverage the existing explainability libraries, which arguably are written for Python.
+
+## Proposal
+
+### Proposal 1 - Embedding: Java -> Python bridge
+
+[jep](https://github.com/ninia/jep) is a Java library that allows embedding Python code in Java applications.
+For this proposal:
+
+* An API wrapper, written in Python will be used as a “front-end” to the Python libraries. This API will have the responsibility of translating Java data structures into Python and vice versa.
+* The Java service will execute the Python algorithms by invoking the API wrapper and passing all the necessary data, read from storage.
+* Python results will be converted into Java data structures
+* Results will be returned by the service (either as responses to ad-hoc requests, or as a Prometheus data). This step is handled exclusively by the Java part. As such, the external communication is delegated to the service.
+
+The API wrapper would provide a general enough abstraction for explainers and metrics so that they could be treated by the TrustyAI service as a regular Java implementation. Any Python implementation-specific initialization or configuration would be implemented on the wrapper side.
+jep provides specific optimizations for the serialization/deserialization of numpy data (in the form of Java types `NDFloat`, etc) which could be leveraged in this scenario.
+
+![](assets/ADR-0001-a.png)
+
+**Pros:**
+
+* External algorithms co-located with TrustyAI service
+* Fewer external calls / network communication
+
+**Cons:**
+
+* Increase complexity of TrustyAI service (dependency tracking and updating)
+* Adding new algorithms requires hardcoding
+* Reduced flexibility (third party libraries need to be added directly)
```

Comment:
I think these two cons can be mitigated with some strategy to "customise" the image and let TrustyAI Service discover the algorithms (via ServiceLoader for instance) available during the startup but it has some complexity that we should properly analyse.
I agree to keep them as cons but I would probably mention the complexity about how to "assemble"/compose the image of the Service more then the hardcoding aspect.
Wdyt?

Reply:
I agree. I've added more emphasis to the composition and that hardcoding would need to happen if not using custom images/composition.